[FLINK-14825][state-processor-api][docs] Rework state processor api documentation

sjwiesman opened the pull request:
## What is the purpose of the change

Update the state processor api documentation based on user feedback to clarify usage and examples. The introduction is taken from the state proc blog post on the flink blog[1]. 

[1] https://flink.apache.org/feature/2019/09/13/state-processor-api.html

sjwiesman commented:
cc @NicoK

qinjunjerry commented on docs/dev/libs/state_processor_api.md:
```diff
@@ -23,166 +23,99 @@ specific language governing permissions and limitations
 </div>
 
-When reading operator state, simply specify the operator uid, state name, and type information.
+### Operator State
```
I think adding the following generalized sentence here before expanding to different kind of operator states is helpful, especially for understanding those parameters given in the followed examples:

>When reading operator state, users specify the operator uid, the state name, and the type information.

qinjunjerry commented on docs/dev/libs/state_processor_api.md:
```diff
@@ -229,6 +206,8 @@ val listState = savepoint.readListState(
 </div>
 </div>
 
+### Keyed State
```
Similar here:
> When reading a keyed state, users specify the operator id and a KeyedStateReaderFunction. Using KeyedStateReaderFunction allows reading...

qinjunjerry commented on docs/dev/libs/state_processor_api.md:
```diff
@@ -333,20 +318,165 @@ val keyedState = savepoint.readKeyedState("my-uid", new ReaderFunction());
 
-State writers are based around the abstraction of `Savepoint`, where one `Savepoint` may have many operators and the state for any particular operator is created using a `BootstrapTransformation`.
+`Savepoint`'s may also be written, which allows such use cases as bootstrapping state based on historical data.
+Each savepoint is made up of many `BootstrapTransformation`'s (explained below), which define the state for an individual operator.
```
Maybe this?
>... Each savepoint is made up of some `BootstrapTransformation`'s (explained below), each of which defines the state for an individual operator.

qinjunjerry commented:
Other than the comments above, it was a good reading: more clear with the pictures :)

sjwiesman commented:
@qinjunjerry Thanks for taking a look. I believe I have addressed all your comments if you'd like to take another pass.

qinjunjerry commented:
Thanks for the update. All looks good to me :)

NicoK commented on docs/dev/libs/state_processor_api.md:
```diff
@@ -333,20 +325,165 @@ val keyedState = savepoint.readKeyedState("my-uid", new ReaderFunction());
+
+    @throws[Exception]
+    override processElement(value: CurrencyRate, ctx: Context): Unit = {
+        ctx.getBroadcastState(descriptor).put(value.currency, value.rate)
```
Can you add the `descriptor` definition somewhere?

NicoK commented on docs/dev/libs/state_processor_api.md:
```diff
@@ -333,20 +325,165 @@ val keyedState = savepoint.readKeyedState("my-uid", new ReaderFunction());
+
+    @Override
+    public void processElement(CurrencyRate value, Context ctx) throws Exception {
+        ctx.getBroadcastState(descriptor).put(value.currency, value.rate);
```
Can you add the `descriptor` definition somewhere?

NicoK commented on docs/dev/libs/state_processor_api.md:
```diff
@@ -413,45 +551,25 @@ The `KeyedStateBootstrapFunction` supports setting event time and processing tim
 The timers will not fire inside the bootstrap function and only become active once restored within a `DataStream` application.
 If a processing time timer is set but the state is not restored until after that time has passed, the timer will fire immediatly upon start.
```
```suggestion
If a processing time timer is set but the state is not restored until after that time has passed, the timer will fire immediately upon start.
```

NicoK commented on docs/dev/libs/state_processor_api.md:
```diff
@@ -413,45 +551,25 @@ The `KeyedStateBootstrapFunction` supports setting event time and processing tim
-</div>
-</div>
-		
 Besides creating a savepoint from scratch, you can base on off an existing savepoint such as when bootstrapping a single new operator for an existing job.
```
```suggestion
Besides creating a savepoint from scratch, you can base one off an existing savepoint such as when bootstrapping a single new operator for an existing job.
```

NicoK commented on docs/dev/libs/state_processor_api.md:
```diff
@@ -413,45 +551,25 @@ The `KeyedStateBootstrapFunction` supports setting event time and processing tim
+    .load(new MemoryStateBackend(), oldPath)
     .withOperator("uid", transformation)
     .write(newPath)
 {% endhighlight %}
```
About potential corruption below: I guess, you could delete the new savepoint though. Can you rephrase that?

sjwiesman replied:
No you cannot delete the new savepoint. It would delete the shared operators.

qinjunjerry replied:
I believe there are three parameters for Savepoint.load(): ExecutionEnvironment, a path and StateBackend. Could you double check? To me, the Savepoint.load() call at the beginning of the "Reading State" section is a good one.

NicoK replied:
Good catch, @qinjunjerry 

@sjwiesman: what I meant is that you can always remove the new savepoint without affecting the old one because the old one is not touched, is it?

NicoK replied:
ok, let's stay cautious here - it depends how you delete the new savepoint: if Flink (in the future) would expose a savepoint delete method, this would then only look at the metadata file and also delete files from the old one, so we keep the sentence as it is

sjwiesman commented on docs/dev/libs/state_processor_api.md:
```diff
@@ -23,166 +23,103 @@ specific language governing permissions and limitations
-A database can be thought of as one or more namespaces, each containing a collection of tables.
-Those tables in turn contain columns whose values have some intrinsic relationship between them, such as being scoped under the same key.
+But let's first have a look at what a stateful Flink job looks like.
+A Flink job is composed of operators, typically one or more source operators, a few operators for the actual processing, and one or more sink operators.
```
I think it should actually be a ';' but yes

sjwiesman commented on docs/dev/libs/state_processor_api.md:
```diff
@@ -23,166 +23,103 @@ specific language governing permissions and limitations
 {% highlight scala %}
 val bEnv      = ExecutionEnvironment.getExecutionEnvironment()
-val savepoint = Savepoint.load(bEnv, "hdfs://path/", new RocksDBStateBackend())
+val savepoint = Savepoint.load(bEnv, "hdfs://path/", new MemoryStateBackend())
```
I've really grown to dislike scala

sjwiesman commented on docs/dev/libs/state_processor_api.md:
```diff
@@ -276,12 +264,13 @@ Then it can read by defining an output type and corresponding KeyedStateReaderFu
-class KeyedState {
-  Integer key;
-  Integer value;
+public class KeyedState {
```
👍

sjwiesman commented:
@NicoK I believe I have addressed all your comments except: 
1) I did not update the picture to change `v1`, `v2` to another letter. I did not create the graphic so I will need to find the person who did. I'm not sure how important you think that is. 

2) Async I/O functions are not stateful so I'm not sure there's anything to say there

3) I'd prefer to just add window documentation in the branch with that feature

qinjunjerry commented on docs/dev/libs/state_processor_api.md:
```diff
@@ -239,114 +225,302 @@ public class StatefulFunctionWithTime extends KeyedProcessFunction<Integer, Inte
+</div>
+</div>
+
+The [UIDs]({{ site.baseurl}}/ops/state/savepoints.html#assigning-operator-ids) associated with each operator much match one to one with the UIDs assigned to the operators in your `DataStream` application; these are how Flink knows what state maps to which operator.
```
Typo: ... each operator **much** match ... -> ... each operator much **must** ...

sjwiesman replied:
Nice catch!

sjwiesman commented:
cc @NicoK
